docs/design/jet/023-jobupload.md

```diff
@@ -0,0 +1,164 @@
+|--------------------------------|--------------------------------------------------|
+| Related Jira                   | _https://hazelcast.atlassian.net/browse/HZ-1455_ |
+| Related Github issues          | _None_                                           |
+| Document Status / Completeness | DRAFT                                            |
+| Requirement owner              | _TBD_                                            |
+| Developer(s)                   | _Orçun Çolak_                                    |
+| Quality Engineer               | _TBD_                                            |
+| Support Engineer               | _TBD_                                            |
+| Technical Reviewers            | _Frantisek Hartman_                              |
+| Simulator or Soak Test PR(s)   | _TBD_                                            |
+
+### Background
+#### Description
+
+- Currently, the job upload can only be performed by hz-cli command. This command requires a JVM on the host machine. The purpose of this feature is to enable non-java clients to be able to upload jobs to Jet server 
+
+
+#### Terminology
+
+| Term                   | Definition                                                              |
+|------------------------|-------------------------------------------------------------------------|
+| Client Binary Protocol | The binary messages used between Hazelcast Clients and Hazelcast Server |
+
+#### Actors and Scenarios
+
+Any language that implements Client Binary Protocol
+
+### Functional Design
+#### Summary of Functionality
+
+Provide a list of functions user(s) can perform.
+- Upload a jar file and pass the same parameters that can be used with hz-cli
+
+#### Additional Functional Design Topics
+
+None
+
+##### Notes/Questions/Issues
+
+- The new functionality requires that all resources accessed by the uploaded job to be available on the server. For example a text file that is used to populate an IMap needs to be available on the server or within the uploaded jar. Because a server side job can not access client side resources.  
+- The client should send multi parts in a sequential manner. Out of order messages are not handled since it would require more resources on the server.
+  
+
+### User Interaction
+#### API design and/or Prototypes
+
+```java
+HazelcastInstance client = ...;
+JetService jetService = client.getJet();
+List<String> jobParameters = emptyList();
+
+String job1 = "job1";
+//If there is an error, throws JetException
+jetService.submitJobJar(getJarPath(),
+        null,
+        job1,
+        null,
+        jobParameters);
+```
+
+#### Client Related Changes
+A new method has been added to JetService interface. Any client that implements this interface, has to implement this method
+```java
+void submitJobJar(@Nonnull Path jarPath, String snapshotName, String jobName, String mainClass, List<String> jobParameters);
```

Comment:
naming nit: maybe `submitJobFromJar`? `submitJobJar` sounds to me like we are only submitting jar to the cluster, `submitJobFromJar` is more around actual job being scheduled

Reply:
Ok I will rename the method and TDD as submitJobFromJar